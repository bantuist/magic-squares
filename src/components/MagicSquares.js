import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setGridSize, updateGrid } from '../actions/magicSquares';
import Board from './Board';
import Controls from './Controls';
import styled from 'styled-components';

const Container = styled.div``;

export class MagicSquares extends Component {
  clue = () => {
    const { cells, solution } = this.props;

    const cellId = Object.keys(cells).find((key, i) => {
      // Set correct cell value if cell is empty or has the wrong value
      return !cells[key].value || cells[key].value !== solution[i]
    });

    
    if (cellId) {
      this.props.updateGrid(
        cellId,
        solution[cellId],
        null
      );
    }
  }
  solve = () => {
    const { cells, solution } = this.props;

    Object.keys(cells).forEach((key, i) => {
      const cell = cells[key];

      this.props.updateGrid(
        cell.id,
        solution[i],
        null
      );
    });
  }
  reset = () => {
    const { cells } = this.props;
  
    Object.keys(cells).forEach((key, i) => {
      const cell = cells[key];

      this.props.updateGrid(
        cell.id,
        0,
        null
      );
    });
  }

  render() {
    const { 
      gridSize, 
      cells, 
      totals,
      total,
      updateGrid,
      activeElement,
    } = this.props;

    return (
      <Container >
        <Board 
          gridSize={gridSize} 
          cells={cells}
          totals={totals}
          boardTotal={total}
          onUpdateGrid={updateGrid}
          activeElement={activeElement}
        />
        <Controls 
          onSolve={this.solve}
          onClue={this.clue}
          onReset={this.reset}
        />
      </Container>
    );
  }
}

export default connect(({ gridSize, total, solution, cells, totals, activeElement }) => {
  return { gridSize, total, solution, cells, totals, activeElement };
}, { setGridSize, updateGrid })(MagicSquares);