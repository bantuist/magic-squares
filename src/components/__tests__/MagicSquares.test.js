import React from 'react';
import { shallow } from 'enzyme';
import { MagicSquares } from '../MagicSquares';

describe('MagicSquares', () => {
  const magicSquares = shallow(<MagicSquares />);

  it('renders without crashing', () => {
    expect(magicSquares).toMatchSnapshot();
  });
});
  