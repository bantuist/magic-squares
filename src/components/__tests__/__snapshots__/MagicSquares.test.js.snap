// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MagicSquares renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MagicSquares />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Board
          activeElement={undefined}
          boardTotal={undefined}
          cells={undefined}
          gridSize={undefined}
          onUpdateGrid={undefined}
          totals={undefined}
/>,
        <Controls
          onClue={[Function]}
          onReset={[Function]}
          onSolve={[Function]}
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "function",
        "props": Object {
          "activeElement": undefined,
          "boardTotal": undefined,
          "cells": undefined,
          "gridSize": undefined,
          "onUpdateGrid": undefined,
          "totals": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "function",
        "props": Object {
          "onClue": [Function],
          "onReset": [Function],
          "onSolve": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Board
            activeElement={undefined}
            boardTotal={undefined}
            cells={undefined}
            gridSize={undefined}
            onUpdateGrid={undefined}
            totals={undefined}
/>,
          <Controls
            onClue={[Function]}
            onReset={[Function]}
            onSolve={[Function]}
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "function",
          "props": Object {
            "activeElement": undefined,
            "boardTotal": undefined,
            "cells": undefined,
            "gridSize": undefined,
            "onUpdateGrid": undefined,
            "totals": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "function",
          "props": Object {
            "onClue": [Function],
            "onReset": [Function],
            "onSolve": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
